### DISCUSSION TITLE
Slow commit causes foreign key error in microservice

### DISCUSSION THREAD

**Discussion by upeshjindal** [fastapi/fastapi#6442] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
async def get_database():
    async with AsyncSessionLocal() as session:
        try:
            yield session
            await session.commit()
            logger.info("Finished commit")
        finally:
            await session.close()
```

### Description

I have two microservices.

1. OrderingService
2. CustomerService

OrderingService checks with CustomerService if the delivery address is present or not. If not, then it calls it to create one. Then uses this address as a ForeignKey in the order.

This works randomly fine. I see it's highly dependent on the speed with which database can write to disk when we commit. If OrderingService is faster than this write-to-disk, it fails.

I'm using SQLAlchemy async. I'm wondering if I'm doing something wrong with calling a service from another service and using its response as ForeginKey.

### Operating System

macOS

### Operating System Details

Postgres 14
SQLAlchemy 1.4.26

### FastAPI Version

0.70.0

### Python Version

3.10

**Comment by oligond** [comment]
> Then uses this address as a ForeignKey in the order.

I have to admit this sounds a little weird to me. Maybe you can expand your example to further clarify this.

It also doesn't seem to be the best idea to have separate, but functionally "bundled" HTTP requests depend on each other in this fashion, especially in a microservicy environment.

Maybe the root cause for your problem is a architectural/design issue rather than related to SQLAlchemy or FastAPI specifically?

But that's surely only some random guy's wild guesses based on the very thin ice of little context information. 😉

**Comment by mayteio** [comment]
How does CustomerService know which address to create if it doesn't exist?

FWIW I agree with @oligond, using an address as a foreign key doesn't sound right.

**Comment by upeshjindal** [comment, marked as answer]
@oligond and @mayteio Thanks for your inputs! I figured what the problem was. The commit() call in the dependency was misplaced. I forgot moving it out while refactoring.

commit is now done when all the work is done and just before sending the response out.

The flow is:

1. Ordering service gets full details of the order including customer data/address info and order items.
2. If the provided customer and address don't exist, create them using CustomerService. 
3. When I got the address back, now I can set the delivery address as a FK in the order. I see you suggest that I don't do this. May I know how else you would connect an order with its delivery address.

Problem was happening at #2: This is where the commit() was happening after I got the response back to OrderingService. The response had the newly created address but not written to disk yet.
